### PR TITLE
Use import/export API for loading dashboards and index pattern

### PR DESCRIPTION
**Pull request by monicasarbu**
This is trying to implement https://github.com/elastic/beats/issues/4409. 

The Beats 6.x dashboards are different than the Beats 5.x dashboards and are not compatible. Thanks to the Kibana API, we can have a json file for each module instead of multiple json files (for dashboards, visualizations, searches, etc). Thus, the way to import/export dashboards differs from the Kibana version that should be equal to the Elasticsearch version.

The json file containing the index-pattern contain as `version` the Beat version, and the dashboards contain as `version` the Kibana version that exported the dashboards.

If Kibana version is 5.x and older: 
- The dashboards should be imported/exported directly from Elasticsearch, and it works only if Elasticsearch is set as output
- All the Beats dashboards are moved to  the `_meta/kibana/5.x` directory 

If Kibana version is 6.x and greater:
- The dashboards should be imported/exported via the Kibana API, so the user needs to configure the kibana url under the `setup.kibana*` settings.
- All the Beats 6.x dashboards should be added to the `_meta/kibana/default` directory for each Beat module.

In order to make use of the import/export API the following changes are needed:
- [x] move all the old dashboards, search, visualizations, index-pattern to `_meta/kibana/5.x` directory
and create a new directory under `_meta/kibana/default` for the new dashboards and index-pattern that are importable via the Kibana API
- [x] update generate_index_pattern to generate the index-pattern for 5.x and >= 6.x versions
- [x] use the Kibana API to load the dashboards and the index pattern for Kibana >= 6
- [x] update import_dashboards script
- [x] update `make update`
- [x] update integration tests

Note: The older dashboards are moved to the `_meta/kibana/5.x` directory only for Metricbeat for now.

### CODE REVIEW

**Comment by ruflin** [comment]
@monicasarbu Should this also go under `dev-tools/cmd/*` like all the other commands?

**Comment by monicasarbu** [comment]
@ruflin This PR is at a very early stage. My plan is to move the import and export functionality to libbeat and then create a script to use it. For now it's just a script (that I will remove) only for testing the import/export API.

**Comment by tsg** [comment]
I think we need to add `setup.kibana` to the short config as well, using `localhost` by default.

**Comment by monicasarbu** [comment]
@tsg Fixed, thanks!

File: libbeat/dashboards/dashboards.go [hunk 1]
```diff
@@ -26,19 +19,104 @@ func ImportDashboards(beatName, beatVersion string, esClient DashboardLoader, cf
...
+}
+
+func ImportDashboardsViaKibana(config *common.Config, dashConfig *DashboardsConfig) error {
+
+	kibanaLoader, err := NewKibanaLoader(config, dashConfig, nil)
+	if err != nil {
+		return fmt.Errorf("fail to create the Kibana loader: %v", err)
+	}
+
+	if kibanaLoader == nil {
```
Comment (tsg) [inline libbeat/dashboards/dashboards.go]: Seems that etiher `err == nil` or `kibanaLoader != nil`, so I think the second check here is not needed and it doesn't follow our usual error checking pattern.

File: libbeat/dashboards/kibana_loader.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	content, err := ioutil.ReadFile(file)
+	if err != nil {
+		return fmt.Errorf("fail to read index-pattern: %v", err)
+	}
+
+	return loader.client.ImportJSON(importAPI, params, bytes.NewBuffer(content))
+}
+
+func (loader KibanaLoader) Close() error {
+	//return loader.client.Close()
```
Comment (tsg) [inline libbeat/dashboards/kibana_loader.go]: Seems that there's nothing to do on Close. Then you can just delete the comment, not to confuse future readers.

File: libbeat/setup/kibana/client.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	kibanaURL, err := common.MakeURL(config.Protocol, config.Path, config.Host, 5601)
+	if err != nil {
+		logp.Err("Invalid host param set: %s, Error: %v", config.Host, err)
+		return nil, fmt.Errorf("invalid Kibana host: %v", err)
+	}
+
+	logp.Debug("kibana", "Kibana url: %s", kibanaURL)
+
+	client := &Client{
+
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: Minor issue, but this extra empty line could be deleted.

File: libbeat/setup/kibana/client.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		Connection: Connection{
+			URL: kibanaURL,
+			http: &http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // ignore expired SSL certificates
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: Hmm, does this mean that HTTPS is always enabled? The comment looks like "insecure by default", we have to be careful with this.

File: libbeat/dashboards/kibana_loader.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	config       *DashboardsConfig
+	version      string
+	msgOutputter *MessageOutputter
+}
+
+func NewKibanaLoader(cfg *common.Config, dashboardsConfig *DashboardsConfig, msgOutputter *MessageOutputter) (*KibanaLoader, error) {
+
+	if cfg == nil {
+		cfg = common.NewConfig()
+		err := cfg.SetBool("enabled", -1, true)
```
Comment (tsg) [inline libbeat/dashboards/kibana_loader.go]: "enabled" is forced to true here, but I didn't see where it's actually checked. We should make sure that disabling with `enabled: false` works.

File: libbeat/setup/kibana/client.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+func NewKibanaClient(cfg *common.Config) (*Client, error) {
+
+	config := defaultKibanaConfig
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, err
+	}
+
+	kibanaURL, err := common.MakeURL(config.Protocol, config.Path, config.Host, 5601)
+	if err != nil {
+		logp.Err("Invalid host param set: %s, Error: %v", config.Host, err)
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: We generally avoid `logp.Err` immediately followed by returning an error.

File: libbeat/setup/kibana/client.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		},
+	}
+	return client, nil
+}
+
+func (conn *Connection) Request(method, extraPath string, params url.Values, body io.Reader) (int, []byte, error) {
+
+	reqURL, err := addPath(conn.URL, extraPath)
+	if err != nil {
+		return 0, nil, fmt.Errorf("fail to parse URL %s: %v", conn.URL, err)
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: Would be perhaps better to write the "fail to parse URL" error in the `addPath` method?

File: libbeat/setup/kibana/client.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	logp.Debug("kibana", "HTTP request URL: %s", reqURL)
+
+	req, err := http.NewRequest(method, reqURL, body)
+	if err != nil {
+		logp.Err("Fail to create a HTTP %s request: %v", method, err)
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: Return error instead of logging.

File: libbeat/setup/kibana/client.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		return 0, nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := conn.http.Do(req)
+	if err != nil {
+		logp.Err("Fail to execute a HTTP %s request: %v", method, err)
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: same here.

File: libbeat/setup/kibana/client.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	type kibanaVersionResponse struct {
+		Name    string `config:"name"`
+		Version struct {
+			Number string `config:"number"`
+		} `config:"version"`
+	}
+
+	_, result, err := client.Connection.Request("GET", "/api/status", nil, nil)
+	if err != nil {
+		return "", fmt.Errorf("HTTP GET request to /api/status fails: %v", err)
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: Please add the `result` to the error message.

File: libbeat/setup/kibana/client.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	var kibanaVersion kibanaVersionResponse
+	err = json.Unmarshal(result, &kibanaVersion)
+	if err != nil {
+		return "", fmt.Errorf("fail to unmarshall the HTTP response from Kibana %s: %v", client.Connection.URL, err)
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: `unmarshal` with one `l`.

File: libbeat/setup/kibana/client.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	return kibanaVersion.Version.Number, nil
+}
+
+func (client *Client) ImportJSON(url string, params url.Values, body io.Reader) error {
+
+	statusCode, response, err := client.Connection.Request("POST", url, params, body)
+	if err != nil {
+		return fmt.Errorf("%v. Response: %s", err, response)
+	}
+	if statusCode != 200 {
```
Comment (tsg) [inline libbeat/setup/kibana/client.go]: 200 might be too strict here. Maybe `> 300`?

File: libbeat/dashboards/kibana_loader.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+	"io/ioutil"
+	"net/url"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/setup/kibana"
+)
+
+var importAPI = "/api/kibana/dashboards/import"
+var exportAPI = "/api/kibana/dashboards/export"
```
Comment (tsg) [inline libbeat/dashboards/kibana_loader.go]: `exportAPI` doesn't seem to be used.

File: libbeat/scripts/Makefile [hunk 14]
```diff
@@ -305,16 +306,8 @@ index-template: ## @build Generate index templates for the given $VERSION. This
 ### KIBANA FILES HANDLING ###
 ES_URL?=http://localhost:9200
 
-.PHONY: export-dashboards
-export-dashboards: python-env update
-	. ${PYTHON_ENV}/bin/activate && python ${ES_BEATS}/dev-tools/export_dashboards.py --url ${ES_URL} --dir $(shell pwd)/_meta/kibana --regex ${BEAT_NAME}-*
```
Comment (tsg) [inline libbeat/scripts/Makefile]: Could we get `export-dashboards` and `import-dashboards` targets based on the go code? I'm good with follow up PR for them.

File: libbeat/setup/kibana/client.go (hunk 4 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: This is something that we need to address in the next PR.

File: libbeat/setup/kibana/client.go (hunk 7 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: fixed

File: libbeat/setup/kibana/client.go (hunk 6 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: fixed

File: libbeat/dashboards/kibana_loader.go (hunk 5 above)
Comment (monicasarbu) [inline libbeat/dashboards/kibana_loader.go]: fixed

File: libbeat/setup/kibana/client.go (hunk 3 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: fixed

File: libbeat/dashboards/kibana_loader.go (hunk 2 above)
Comment (monicasarbu) [inline libbeat/dashboards/kibana_loader.go]: fixed

File: libbeat/setup/kibana/client.go (hunk 8 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: fixed

File: libbeat/setup/kibana/client.go (hunk 9 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: fixed

File: libbeat/setup/kibana/client.go (hunk 10 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: fixed

File: libbeat/setup/kibana/client.go (hunk 11 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: fixed

File: libbeat/setup/kibana/client.go (hunk 12 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: true, fixed

File: libbeat/dashboards/kibana_loader.go (hunk 13 above)
Comment (monicasarbu) [inline libbeat/dashboards/kibana_loader.go]: fixed

File: libbeat/scripts/Makefile (hunk 14 above)
Comment (monicasarbu) [inline libbeat/scripts/Makefile]: It's a bit tricky for `export-dashboards` as we need to pass the list of ids that we want to export in one json file. Usually you would run it to export the dashboards for a module, not for a Beat. Added to the TODO list in the meta issue: https://github.com/elastic/beats/issues/4409

File: libbeat/setup/kibana/client.go (hunk 4 above)
Comment (monicasarbu) [inline libbeat/setup/kibana/client.go]: Added to the TODO list: https://github.com/elastic/beats/issues/4409

**Comment by exekias** [comment]
I think it would make sense to add `dashboards` to the existing `export` command, I think users could use that to share easily their own dashboards

**Comment by monicasarbu** [comment]
@ruflin yes, that's the plan to freeze the 5.x dashboards and improve only the 6.x dashboards.

File: .gitignore [hunk 15]
```diff
@@ -22,6 +22,9 @@ coverage.out
 *.swo
 *.swn
 
+# Executable
```
Comment (monicasarbu) [inline .gitignore]: Good idea, done.

File: auditbeat/auditbeat.reference.yml [hunk 16]
```diff
@@ -664,6 +664,25 @@ output.elasticsearch:
...
+
+# Starting with Beats version 6.0.0, the dashboards are loaded via the Kibana API. 
+# This requires a Kibana endpoint configuration.
+setup.kibana:
+
+  # Kibana Host
+  # Scheme and port can be left out and will be set to the default (http and 5601)
+  # In case you specify and additional path, the scheme is required: http://localhost:5601/path
+  # IPv6 addresses should always be defined as: https://[2001:db8::1]:5601
+  # host: "localhost:5601"
```
Comment (monicasarbu) [inline auditbeat/auditbeat.reference.yml]: done!

File: libbeat/dashboards/dashboards.go [hunk 17]
```diff
@@ -26,19 +19,107 @@ func ImportDashboards(beatName, beatVersion string, esClient DashboardLoader, cf
...
+	if err := importer.Import(); err != nil {
+		return false, fmt.Errorf("fail to import the dashboards in Elasticsearch: %v", err)
+	}
+
+	return true, nil
+
+}
+func getMajorVersion(version string) (int, error) {
+
+	majorVersion := string(version[0])
```
Comment (monicasarbu) [inline libbeat/dashboards/dashboards.go]: done

**Comment by tsg** [comment]
jenkins, test it

**Comment by tsg** [comment]
jenkins, test it

**Comment by tsg** [comment]
jenkins, test it

**Pull request merged**